File: src/main/java/org/athenian/java/ThreadSafety.java
package org.athenian.java;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class ThreadSafety {

    public static void main(String... args)
            throws InterruptedException {

        final int JOB_COUNT = 100;
        final int INC_COUNT = 1_000;

        final CountDownLatch latch = new CountDownLatch(JOB_COUNT);
        final NonThreadSafeCounter nonThreadSafeCounter = new NonThreadSafeCounter();
        final SynchronizedMethodCounter synchronizedMethodCounter = new SynchronizedMethodCounter();
        final SynchronizedBlockCounter synchronizedBlockCounter = new SynchronizedBlockCounter();
        final LockCounter lockCounter = new LockCounter();
        final AtomicCounter atomicCounter = new AtomicCounter();

        for (int i = 0; i < JOB_COUNT; i++) {
            new Thread(
                    () -> {
                        for (int j = 0; j < INC_COUNT; j++) {
                            nonThreadSafeCounter.increment();
                            synchronizedMethodCounter.increment();
                            synchronizedBlockCounter.increment();
                            lockCounter.increment();
                            atomicCounter.increment();
                        }
                        // Indicate job is complete
                        latch.countDown();
                    }).start();
        }

        // Wait for all jobs to complete
        latch.await();

        System.out.printf("Non-thread-safe counter value = %d%n", nonThreadSafeCounter.value());
        System.out.printf("Synchronized method counter value = %d%n", synchronizedMethodCounter.value());
        System.out.printf("Synchronized block counter value = %d%n", synchronizedBlockCounter.value());
        System.out.printf("Lock counter value = %d%n", lockCounter.value());
        System.out.printf("Atomic counter value = %d%n", atomicCounter.value());
    }

    static class NonThreadSafeCounter {
        int count = 0;

        void increment() {
            this.count++;
        }

        int value() {
            return this.count;
        }
    }

    static class SynchronizedMethodCounter {
        int count = 0;

        synchronized void increment() {
            this.count++;
        }

        int value() {
            return this.count;
        }
    }

    static class SynchronizedBlockCounter {
        private final static Object OBJECT = new Object();
        int count = 0;

        void increment() {
            synchronized (this /* OBJECT */) {
                this.count++;
            }
        }

        int value() {
            return this.count;
        }
    }

    static class LockCounter {
        final Lock lock = new ReentrantLock();
        int count = 0;

        void increment() {
            lock.lock();
            try {
                this.count++;
            } finally {
                lock.unlock();
            }
        }

        int value() {
            return this.count;
        }
    }

    static class AtomicCounter {
        final AtomicInteger count = new AtomicInteger(0);

        void increment() {
            this.count.incrementAndGet();
        }

        int value() {
            return this.count.get();
        }
    }
}

File: src/main/kotlin/org/athenian/kotlin/ThreadSafety.kt
package org.athenian.kotlin

import java.util.concurrent.CountDownLatch
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.thread

fun main() {

    val JOB_COUNT = 100
    val INC_COUNT = 1_000

    val latch = CountDownLatch(JOB_COUNT)
    val nonThreadSafeCounter = NonThreadSafeCounter()
    val synchronizedMethodCounter = SynchronizedMethodCounter()
    val synchronizedBlockCounter = SynchronizedBlockCounter()
    val lockCounter = LockCounter()
    val atomicCounter = AtomicCounter()

    repeat(JOB_COUNT) { i ->
        thread {
            repeat(INC_COUNT) {
                nonThreadSafeCounter.increment()
                synchronizedMethodCounter.increment()
                synchronizedBlockCounter.increment()
                lockCounter.increment()
                atomicCounter.increment()
            }
            // Indicate job is complete
            latch.countDown()
        }
    }

    // Wait for all jobs to complete
    latch.await()

    println("Non-thread-safe counter value = ${nonThreadSafeCounter.value()}")
    println("Synchronized method counter value = ${synchronizedMethodCounter.value()}")
    println("Synchronized block counter value = ${synchronizedBlockCounter.value()}")
    println("Lock counter value = ${lockCounter.value()}")
    println("Atomic counter value = ${atomicCounter.value()}")
}

class NonThreadSafeCounter {
    var count = 0

    fun increment() {
        this.count++
    }

    fun value() = count
}

class SynchronizedMethodCounter {
    var count = 0

    @Synchronized
    fun increment() {
        this.count++
    }

    fun value() = count
}

class SynchronizedBlockCounter {
    var count = 0

    fun increment() {
        synchronized(this /* OBJECT */) {
            count++
        }
    }

    fun value(): Int {
        return count
    }

    companion object {
        private val OBJECT = Any()
    }
}

class LockCounter {
    var count = 0
    val lock = ReentrantLock()

    fun increment() {
        lock.lock()
        try {
            count++
        } finally {
            lock.unlock()
        }
    }

    fun value() = count
}

class AtomicCounter {
    val count = AtomicInteger(0)

    fun increment() {
        count.incrementAndGet()
    }

    fun value() = count.get()
}